### PR TITLE
Singleton edits

### PR DESCRIPTION
**Pull request by EricJIsaac**
Hey Spencer, I edited the singleton behaviour slightly & added a git ignore file.

### CODE REVIEW

**Comment by spencerturkel** [comment]
Disagree with how you changed the behavior - a dynamically allocated AppServer with pointer semantics rather than reference semantics leaves room for bugs in the other guy's code. When exactly do they call shutdown? Who calls shutdown? It SHOULD be deleted at the end of the program BEFORE the DB resource is released so that we can save all the cache data - and the best way to enforce that is to use the reference semantics like I began with so that this is done automatically in the destructor and the other guys don't even have to know about how it works. No shutdown() needed !If I've overlooked something on this point though, please let me know.

Additionally, this means they can write "appserver.blah();" instead of having to write "appserver->blah();" all over the place which is both annoying AND easy to forget. 

Also, AppServer conceptually shouldn't have a copy ctor or assignment, since it is encapsulating a global state - makes no sense to copy a global state.

Besides those issues a .gitignore file is a good idea and so is your commenting, so get rid of the bad stuff and I'll merge this in.

**Comment by EricJIsaac** [comment]
tldr: The edits better reflect how singleton are normally implemented

Now to address each point;
1. Bugs in the other guys code:
 imho it seemed more intuitive to be explicit about when the app server should be shut down, as it is, the app server can easily go out of scope, causing deletion, especially in a GUI application with a somewhat indirect logical flow. In this case, the program would throw an error when attempting an operation on a garbage-collected app server.
2.Who should call delete?
Anyone who wants to shut-down the current instance of the app server (perhaps when it isn't being used). Shutdown also doesn't necessarily need to close the DB, but if we are worried, an instance will be created as soon as it is referenced again.
3. So the other guys don't need to know how it works?
They don't need to know how instance works either. They simply reference it, and call shutdown if they wish to end the current server connection.
4. Destructor public?
You are right on this, no particular reason to have it private, but since no one should be instantiating an App Server, it seemed like good form to be consistent, but there is no issue with it being public.
5. "app server.blah()" vs "app server-> blah()"
The goal was to pass a reference around to the other components that needs to reference the app server. Afaik no other components will need to perform operations on the app server itself, so they need at most a pointer to the current app server to access methods. Also, given the object-oriented nature of C++, I don't think "something-> function" is exceptionally bizarre syntax.
6.Dont need a copy/assignment constructor?
If you don't define these, the compiler will create them for you (how objects get constructed when constructors are not defined), so as it is, you DO have a copy and assignment constructor and they are public. I defined them privately to enforce exactly what you are saying.
7. Unneeded headers?
Afaik, the standard library  and UNIX API's are pre-compiled, and will not increase our compilation times (unless you plan to edit them).
8. Less functions in the global namespace?
no name spaces are being used, so currently no functions are added to the global namespace. You would still need to reference the specific namespace (e.g. std:: string)
9. Better reflects singleton behavior
Similar to Push & Pop for stacks/queues, Instance is generally what the handle to a singleton is named. It's arbitrary, but if any of the others are familiar with singletons, this is probably the method they will look for.

**Comment by spencerturkel** [comment]
1. I just did a bit of testing on this point. Turns out when the reference goes out of scope, all that happens is that the reference itself is destroyed - not the referent object. Our referent object has static storage duration and remains alive from program start to program end.

2-3. Given what I just found with testing, the singleton is always alive right up until termination, so it doesn't make sense to manually shutdown - if the server is unused somewhere then there's no need to grab a reference to it , it will just sit in the static memory allocated at program start, and it will be cleaned up at termination by the dtor. I can't think of a reason why we would need to reset the AppServer...if I'm wrong about this and we DO need to, we can create a reset function.

4. While I was testing I actually tried making the dtor private like you originally did and it works properly, so you were right here, I'll change this back.

I see your points on the rest of everything else! Will change accordingly.
